AMI_Cleanup.py

```diff
@@ -0,0 +1,84 @@
+import boto3
+import shutil
+import datetime
+import dateutil.relativedelta
+import xml.etree.ElementTree as ET
+import subprocess
+import pytz
+
+#TODO: Jenkins is set up to use Justin IAM - change to beambox
+#This script needs to be run on the main jenkins server so that it can update the cloud AMI to use the latest
+#If there is ever a programmatic way to update the cloud config then this could be moved to a lambda
+#Until then it should be setup as a cron job per below - to be executed once every 2 days
+# sudo crontab -e
+# */* * 2 * * /home/ubuntu/AMI_Cleanup.py >> /home/ubuntu/AMI_Cleanup.log 2>&1
+
+utc=pytz.UTC
+
+def getEC2For(region):
+    ec2 = boto3.client('ec2', region_name=region)
+    return ec2
+
+def backup_and_retrieve_config():
+    today = datetime.datetime.today()
+    shutil.copy('/var/lib/jenkins/config.xml','/var/lib/jenkins/config.xml.bak_'+str(today.month)+'_'+str(today.day))
+    return ET.parse('/var/lib/jenkins/config.xml')
+
+def getTaggedAMIsAlongWithMostRecent(region):
+    ec2 = getEC2For(region)
+    amis = ec2.describe_images(Owners=['self'])
+    allAMIs = []
+    newestAMI = None
+    current_date_time = utc.localize(datetime.datetime.now())
+    for ami in amis['Images']:
+        if newestAMI is None:
+            newestAMI = ami
+        if 'beam-automation-' in ami['Name']:
+            ami_creation = dateutil.parser.parse(ami['CreationDate'])
+            current_newest_creation = dateutil.parser.parse(newestAMI['CreationDate'])
+            two_weeks_ago = current_date_time - dateutil.relativedelta.relativedelta(weeks=2)
+            if ami_creation < two_weeks_ago:
+                allAMIs.append((ami, True)) #mark for deletion
+            else:
+                allAMIs.append((ami, False)) #mark for keeping
+            if ami_creation > current_newest_creation:
+                newestAMI = ami
+    return (allAMIs, newestAMI)
+
+def replace_old_ami_with_new_using(oldAMIIds, newestAMIId, config):
+    currentConfigAMIs = config.findall('.//hudson.plugins.ec2.SlaveTemplate/ami')
+    for configAMI in currentConfigAMIs:
+        if configAMI.text in oldAMIIds:
+            configAMI.text = newestAMIId
+    config.write('/var/lib/jenkins/config.xml')
+
+def restart_jenkins():
+    process = subprocess.call(['sudo', 'service', 'jenkins', 'restart'])
+
+def delete_older_amis_and_snapshots(allAMIsTagged, region):
+    amisToKeep = map(lambda x: x[0]['ImageId'], filter(lambda x: x[1] == False, allAMIsTagged))
+    amisAndSnapshotsToDelete = map(lambda x: (x[0]['ImageId'], map(lambda y: y['Ebs']['SnapshotId'], x[0]['BlockDeviceMappings'])), filter(lambda x: x[1] == True, allAMIsTagged))
+    print('Keeping AMI IDs:' + str(amisToKeep))
+    print('Deleting AMI IDs and correlated SnapshotIds for region '++':' + str(amisAndSnapshotsToDelete))
+    ec2 = getEC2For(region)
+    for amiIdAndSnapshotIds in amisAndSnapshotsToDelete:
```

Comment:
This is the most dangerous part - however it SHOULD be safeguarded to not obliterate everything by making sure there are at least 2 AMIs to keep (in the `main` method - maybe it should be moved into here?)

Reply:
The method name is self-explanatory, I think it is fine to leave as is because it is already guarded.